Sage/carbon
# Fe 5505 add forwardref

### Proposed behaviour

All components which wrap a single input should use `React.forwardRef` to allow consumers to pass a `ref={someRef}` to the component and have that be attached to the underlying DOM element. (It should work with any React ref - so either an object with a `current` property or a function which accepts an HTML element.)

The `inputRef` prop which exists on most such elements and which offers similar functionality should be deprecated as it is no longer needed when `ref` works.

Components such as DateRange and NumeralDate which consist of more than one input should offer props to allow references to be taken to each individual input element as needed.

Fix #5564 



### Current behaviour

Most such components do not allow a `ref` to be attached. In most cases this can be done with the `inputRef` function prop but this has a non-standard shape and isn't obvious (most React developers know how to use `ref` and will expect it to work, whereas knowing about `inputRef` requires a careful reading of the docs - in most cases it only appears once in a big props table towards the bottom of the page).



### Checklist




#### QA


### Additional context



Note that I have not removed the `inputRef` prop from the internal `Input` component as the way this is currently implemented, a non-memoised callback can be passed and will not get called with `null` until the component unmounts - whereas if this is replaced with a "genuine" callback ref, a function that changes every render causes the ref to become "detached" and be called with null before being called again with the DOM element. While not a hard issue to fix, this would count as a breaking change - so I have left this as is, and can be removed later when we finally remove the `inputRef` prop from all components.

### Testing instructions



Careful regression tests are needed on all the following components which have been changed.

Checkbox
Date
DateRange
Decimal
FilterableSelect
GroupedCharacter
MultiSelect
Number
NumeralDate
RadioButton
Search
Select
Switch
Textarea
Textbox
TileSelect

Check that the new props work as intended on DateRange (startRef and endRef) and NumeralDate (dayRef, monthRef, yearRef), and that a plain `ref` works on all the others.

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.
